src/bin/tikv-ctl.rs

```diff
@@ -182,3 +313,113 @@ fn dump_range(db: DB, from: String, to: Option<String>, limit: Option<u64>, cf:
                  })
         .unwrap()
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::str;
+    use rocksdb::Writable;
+    use tikv::util::codec::bytes::encode_bytes;
+    use tikv::util::codec::number::NumberEncoder;
+    use tikv::raftstore::store::keys;
+    use tikv::storage::{ALL_CFS, CF_LOCK, CF_WRITE, CF_DEFAULT};
+    use tikv::storage::mvcc::{Lock, Write, LockType, WriteType};
+    use tempdir::TempDir;
+    use tikv::util::rocksdb::new_engine;
+    use tikv::storage::types::Key;
+
+    const PREFIX: &'static [u8] = b"k";
+
+    #[test]
+    fn test_ctl_mvcc() {
+        let tmp_dir = TempDir::new("mvcc_tmp").expect("create mvcc_tmp dir");
+        let file_path = tmp_dir.path().join("tmp_db");
+        let db = new_engine(file_path.to_str().unwrap(), ALL_CFS).unwrap();
+        let test_data: Vec<(_, _, u64)> =
+            vec![(PREFIX, b"v", 5), (PREFIX, b"x", 10), (PREFIX, b"y", 15)];
+        for &(k, v, ts) in &test_data {
+            let key = keys::data_key(Key::from_raw(&k[..]).append_ts(ts).encoded().as_slice());
+            let value = v;
+            db.put(key.as_slice(), &value[..]).unwrap();
+        }
+        let kvs_gen: Vec<MvccKv<Vec<u8>>> = gen_mvcc_iter(&db, "k", CF_DEFAULT).unwrap();
+        let mut test_iter = test_data.clone();
+        assert_eq!(test_iter.len(), test_data.len());
+        for kv in kvs_gen {
+            let ts = kv.key.decode_ts().unwrap();
+            let key = kv.key.truncate_ts().unwrap().raw().unwrap();
+            let value = kv.value.as_slice();
+            test_iter.retain(|&s| !(&s.0[..] == key.as_slice() && &s.1[..] == value && s.2 == ts));
```

Comment:
Why? For example, kvs_gen is supposed to seek k1, k2, k3. But for some bug, it seek k1, k2, k3 and j5. How is this situation detected?